### PR TITLE
Added ResampleLinear() to FloatPoint2DVectorList.

### PR DESCRIPTION
**Pull request by niklasborwell**
Added a function to re-sample a data set with linear interpolation.

### CODE REVIEW

**Comment by niklasborwell** [comment]
@bettinaselig Can you peer review? If you're not too busy

**Comment by bettinaselig** [comment]
Sure.

**Comment by bettinaselig** [comment]
What shall this function do?

**Comment by niklasborwell** [comment]
You mean what it will be used for? Or what the function does?

**Comment by bettinaselig** [comment]
Both.

**Comment by niklasborwell** [comment]
The function is just a simple re-sampling with linear interpolation, no anti-aliasing, but we might extend it later. 
We need this when we try to estimate the sync between the gyro data and the video stream. The problem is that the gyro data has a higher sampling rate, and neither of the sample rates are constant, which means we can't calculate the cross-correlation directly. 
So I will use this to re-sample the gyro and video stream data so that we can compute the cross-correlation.

**Comment by bettinaselig** [comment]
Looks good.

File: source/math/FloatPoint2DVectorList.ooc [hunk 1]
```diff
@@ -107,4 +108,30 @@ FloatPoint2DVectorList: class extends VectorList<FloatPoint2D> {
...
 		result
 	}
+	resampleLinear: func (start, end, interval: Float) -> This {
+		// Assumes list is sorted by x values.
+		resultCount := ((end - start) / interval) ceil() as Int + 1
+		result := This new()
+		previousIndex := 0
+		for (i in 0 .. resultCount) {
+			point := FloatPoint2D new(start + i * interval, 0.0f)
+			if (point x <= this[0] x)
```
Comment (marcusnaslund) [inline source/math/FloatPoint2DVectorList.ooc]: To avoid all the `memcpy`, maybe do the `thisPointer` trick here as well? If you care about performance.

File: source/math/FloatPoint2DVectorList.ooc [hunk 2]
```diff
@@ -107,4 +108,30 @@ FloatPoint2DVectorList: class extends VectorList<FloatPoint2D> {
...
+				point y = this[this count - 1] y
+			else {
+				leftPoint, rightPoint: FloatPoint2D
+				for (j in previousIndex .. this count)
+					if (this[j] x <= point x) {
+						leftPoint = this[j]
+						rightPoint = this[Int minimum(j + 1, this count - 1)]
+						previousIndex = j
+					} else break
+				weight := (Float absolute(point x - leftPoint x) / Float absolute(rightPoint x - leftPoint x)) clamp(0.0f, 1.0f)
```
Comment (marcusnaslund) [inline source/math/FloatPoint2DVectorList.ooc]: Do you really need to `clamp`?

**Comment by marcusnaslund** [comment]
Done reviewing.

File: source/math/FloatPoint2DVectorList.ooc (hunk 2 above)
Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: If `point x > rightPoint x`?

Comment (niklasborwell) [inline source/math/FloatPoint2DVectorList.ooc]: No, you're right, it's not necessary here.

Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: If `point x > rightPoint x`?

File: source/math/FloatPoint2DVectorList.ooc (hunk 1 above)
Comment (niklasborwell) [inline source/math/FloatPoint2DVectorList.ooc]: Well, I thought about it, but I think the difference would be negligible compared to the ugliness of the code, since I'm indexing at so many different indices. Also performance is not critical, so I think we can tackle it if we need it.

File: source/math/FloatPoint2DVectorList.ooc (hunk 2 above)
Comment (niklasborwell) [inline source/math/FloatPoint2DVectorList.ooc]: No, it's covered by the previous if cases.

**Comment by niklasborwell** [comment]
Updated.

File: source/math/FloatPoint2DVectorList.ooc [hunk 3]
```diff
@@ -107,4 +108,30 @@ FloatPoint2DVectorList: class extends VectorList<FloatPoint2D> {
...
+				point y = this[this count - 1] y
+			else {
+				leftPoint, rightPoint: FloatPoint2D
+				for (j in previousIndex .. this count)
+					if (this[j] x <= point x) {
+						leftPoint = this[j]
+						rightPoint = this[Int minimum(j + 1, this count - 1)]
+						previousIndex = j
+					} else break
+				weight := Float absolute(point x - leftPoint x) / Float absolute(rightPoint x - leftPoint x)
```
Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: You never check if `point x` is smaller than `rightPoint x`, so you need to `clamp`. Don't ignore me! I bite you in the ass otherwise!

File: source/math/FloatPoint2DVectorList.ooc (hunk 2 above)
Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: If you say so...

File: source/math/FloatPoint2DVectorList.ooc (hunk 3 above)
Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: See above.

Comment (niklasborwell) [inline source/math/FloatPoint2DVectorList.ooc]: Ok, I double checked this pretty carefully to prove that I'm right! 
It is never checked explicitly, but since we are assuming that the list is sorted by x values, this case won't happen. `rightPoint` is always next in the list after `leftPoint`, so `rightPoint x > leftPoint x`. And the only way `leftPoint x > point x` is if `point x <= this[0] x` which is checked earlier. Also, `rightPoint x == leftPoint x` is never true even if there are duplicate points, because it will skip directly to the second point in the j loop. I will add a test with duplicate points though, that would be useful.

**Comment by niklasborwell** [comment]
Updated.

File: source/math/FloatPoint2DVectorList.ooc (hunk 3 above)
Comment (bettinaselig) [inline source/math/FloatPoint2DVectorList.ooc]: Ok. I am convinced :+1:
